src/Features/CSharp/Portable/ConvertLinq/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,585 @@
+            IEnumerable<StatementSyntax> statementsCannotBeConverted = null;
+
+            void ProcessLocalDeclarationStatement(LocalDeclarationStatementSyntax localDeclarationStatement)
+            {
+                foreach (var variable in localDeclarationStatement.Declaration.Variables)
+                {
+                    convertingNodes.Add(variable);
+                    identifiers.Add(variable.Identifier);
+                }
+            }
+
+            // Setting statementsCannotBeConverted to anything means that we stop processing.
+            while (statementsCannotBeConverted == null)
+            {
+                switch (current.Kind())
+                {
+                    case SyntaxKind.Block:
+                        var block = (BlockSyntax)current;
+                        var array = block.Statements.ToArray();
+                        if (array.Any())
+                        {
+                            // Process all statements except the last one.
+                            for (int i = 0; i < array.Length - 1; i++)
+                            {
+                                var statement = array[i];
+                                if (statement is LocalDeclarationStatementSyntax localDeclarationStatement)
+                                {
+                                    ProcessLocalDeclarationStatement(localDeclarationStatement);
+                                }
+                                else
+                                {
+                                    statementsCannotBeConverted = array.Skip(i);
+                                    break;
+                                }
+                            }
+
+                            // Process the last statement separately.
+                            current = array.Last();
+                        }
+                        else
+                        {
+                            statementsCannotBeConverted = Enumerable.Empty<StatementSyntax>();
+                        }
+                        break;
+
+                    case SyntaxKind.ForEachStatement:
+                        var currentForEachStatement = (ForEachStatementSyntax)current;
+                        identifiers.Add(currentForEachStatement.Identifier);
+                        convertingNodes.Add(currentForEachStatement);
+                        current = currentForEachStatement.Statement;
+                        break;
+
+                    case SyntaxKind.IfStatement:
+                        var ifStatement = (IfStatementSyntax)current;
+                        if (ifStatement.Else == null)
+                        {
+                            convertingNodes.Add(ifStatement);
+                            current = ifStatement.Statement;
+                            break;
+                        }
+                        else
+                        {
+                            statementsCannotBeConverted = new[] { current };
```

Comment:
can we have comments explaining this?  I do not understand why an 'if' cannot be converted if it has an else.